near-sdk/src/types/account_id.rs

```diff
@@ -0,0 +1,200 @@
+
+/// Account identifier. This is the human readable utf8 string which is used internally to index
+/// accounts on the network and their respective state.
+///
+/// Because these IDs have to be validated, they have to be converted from a string
+/// with [`FromStr`] or [`TryFrom`] a compatible type. To skip validation on initialization,
+/// [`AccountId::new_unchecked`] can be used.
+///
+/// # Examples
+/// ```
+/// use near_sdk::AccountId;
+/// use std::convert::{TryFrom, TryInto};
+///
+/// // `FromStr` conversion
+/// let alice: AccountId = "alice.near".parse().unwrap();
+/// assert!("invalid.".parse::<AccountId>().is_err());
+///
+/// let alice_string = "alice".to_string();
+///
+/// // From string with validation
+/// let alice = AccountId::try_from(alice_string.clone()).unwrap();
+/// let alice: AccountId = alice_string.try_into().unwrap();
+///
+/// // Initialize without validating
+/// let alice_unchecked = AccountId::new_unchecked("alice".to_string());
+/// assert_eq!(alice, alice_unchecked);
+/// ```
+///
+/// [`FromStr`]: std::str::FromStr
+#[derive(
+    Debug,
+    Clone,
+    PartialEq,
+    PartialOrd,
+    Ord,
+    Eq,
+    BorshDeserialize,
+    BorshSerialize,
+    Serialize,
+    Hash,
+    BorshSchema,
+)]
+pub struct AccountId(String);
+
+impl AccountId {
+    /// Returns reference to the account ID bytes.
+    pub fn as_bytes(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+    /// Returns reference to the account ID string.
+    pub fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+    /// Consumes `self` to return inner [`String`] representation.
+    pub fn into_string(self) -> String {
+        self.0
+    }
+    /// Constructs new AccountId from `String` without checking validity.
+    pub fn new_unchecked(id: String) -> Self {
+        Self(id)
+    }
+}
+
+impl fmt::Display for AccountId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl From<AccountId> for String {
+    fn from(id: AccountId) -> Self {
+        id.0
+    }
+}
+
+impl AsRef<str> for AccountId {
+    fn as_ref(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+impl<'de> Deserialize<'de> for AccountId {
+    fn deserialize<D>(deserializer: D) -> Result<Self, <D as de::Deserializer<'de>>::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let s: String = Deserialize::deserialize(deserializer)?;
+        Self::try_from(s).map_err(de::Error::custom)
+    }
+}
+
+impl TryFrom<&str> for AccountId {
+    type Error = ParseAccountIdError;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        value.parse()
+    }
+}
+
+impl TryFrom<&[u8]> for AccountId {
+    type Error = ParseAccountIdError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        core::str::from_utf8(value)
+            .map_err(|_| ParseAccountIdError { kind: ParseAccountIdErrorKind::InvalidUtf8 })?
+            .parse()
```

Comment:
`?`-less alternative: 

```suggestion
        core::str::from_utf8(value).map_err(|_| ParseAccountIdError { kind: ParseAccountIdErrorKind::InvalidUtf8 })
            .and_then(str::parse)
```

Reply:
Just for readability?

Reply:
More for minimality, the `?` does early return and `Into` conversion, and we don't really need those here. 

There's also https://github.com/rust-lang/rust/issues/37939, which make me cautious about using `?` where I can cleanly avoid it.